reactor-core/src/main/java/reactor/core/publisher/FluxLimitRequest.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2011-2017 Pivotal Software Inc, All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package reactor.core.publisher;
+
+import java.util.concurrent.atomic.AtomicLongFieldUpdater;
+
+import org.reactivestreams.Subscription;
+import reactor.core.CoreSubscriber;
+
+/**
+ * @author Simon Baslé
+ * @author David Karnok
+ */
+final class FluxLimitRequest<T> extends FluxOperator<T, T> {
+
+	final long cap;
+
+	FluxLimitRequest(Flux<T> flux, long cap) {
+		super(flux);
+		this.cap = cap;
+	}
+
+	@Override
+	public void subscribe(CoreSubscriber<? super T> actual) {
+		source.subscribe(new FluxLimitRequestSubscriber<>(actual, this.cap));
+	}
+
+	@Override
+	public int getPrefetch() {
+		return 0;
+	}
+
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.REQUESTED_FROM_DOWNSTREAM) return cap;
+
+		//FluxOperator defines PREFETCH and PARENT
+		return super.scanUnsafe(key);
+	}
+
+	static class FluxLimitRequestSubscriber<T> implements InnerOperator<T, T> {
+
+		final CoreSubscriber<? super T> actual;
+
+		Subscription parent;
+
+		volatile long toProduce;
```

Comment:
No need for `volatile`. Or at least have a field updater and lazySet updates to it.

Reply:
indeed. removed the `volatile` in latest commit 👍